set_user.c

```diff
@@ -240,10 +285,22 @@ set_user(PG_FUNCTION_ARGS)
 		NewUser_is_superuser = ((Form_pg_authid) GETSTRUCT(roleTup))->rolsuper;
 		ReleaseSysCache(roleTup);
 
-		if (NewUser_is_superuser && Block_SU)
-			ereport(ERROR,
-					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
-					 errmsg("Switching to superuser blocked by set_user config")));
+		if (NewUser_is_superuser)
+		{
+			if (is_privileged && Block_SU)
+				ereport(ERROR,
+						(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						 errmsg("Switching to superuser blocked by set_user config.")));
+			else if (!is_privileged)
+				ereport(ERROR,
+						(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+						 errmsg("Switching to superuser only allowed for privileged procedure: \'set_user_u\'")));
+			else if (!check_user_whitelist(GetUserId(), SU_Whitelist))
```

Comment:
So are we going to require a whitelist? Seems redundant -- if we reject everyone when the whitelist is empty, we can just eliminate Block_SU entirely, can't we? So the empty whitelist is same as "block su" was previously, and when whitelist exists, only those in the list can escalate.